Internal doc for SolverOptions (#2918)
qutip/qutip: doc/internals/solver/solveroptions.rst

.. _internal_solveroptions:

Solver Options Infrastructure
=============================

The configuration of QuTiP's physical solvers and numerical integrators is
handled via specialized dictionary objects. Rather than exposing raw Python
dictionaries directly, the framework utilizes an internal subclass,
:class:`_SolverOptions`.

Motivation
==========

> Simple dictionary or dataclass lacks the feature needed to

Reviewer: ```suggestion
A simple dictionary or dataclass lacks the features needed to:
```

* **Preventing Silent Silent Bugs**: Raw Python dictionaries accept any key-value
  pair without validation. A simple typo, such as typing ``"abstol"`` instead of
  ``"atol"``, would be silently ignored by a standard dictionary, causing the

>   underlying integrator to fall back to default values without ever alerting the user.

Reviewer: ```suggestion
* **Prevent silent bugs**: Raw Python dictionaries accept any key-value
  pair without validation. A simple typo, such as typing ``"abstol"`` instead of
  ``"atol"``, would be silently ignored by a standard dictionary, causing the
  underlying integrator to fall back to default values without ever alerting the user.
```

* **Integrator State Synchronization**: Modern differential equation solvers are
  stateful; they maintain adaptive step histories, interpolation tables, and
  multi-step coefficients. If a user modifies an integration parameter *after*
  a simulation has started stepping forward, the internal state of the underlying
  ODE engine must be forcefully reset to prevent numeric instability or crashes.

>   A standard dictionary cannot broadcast its updates to intercept these mid-run changes.

Reviewer: ```suggestion
* **Integrator state synchronization**: Modern differential equation solvers are
  stateful; they maintain adaptive step histories, interpolation tables, and
  multi-step coefficients. If a user modifies an integration parameter *after*
  a simulation has started stepping forward, the internal state of the underlying
  ODE engine must be forcefully reset to prevent numeric instability or crashes.
  A standard dictionary cannot broadcast its updates to intercept these mid-run changes.
```

* **Namespace Isolation**: Different integration backends support entirely
  different sets of configuration options. For example, a specialized Krylov subspace
  propagator expects parameters like Krylov dimension limits, which are completely

>   meaningless to a standard SciPy ``zvode`` integrator.

Reviewer: ```suggestion
* **Namespace isolation**: Different integration backends support entirely
  different sets of configuration options. For example, a specialized Krylov subspace
  propagator expects parameters like Krylov dimension limits, which are completely
  meaningless to a standard SciPy ``zvode`` integrator.
```

The :class:`_SolverOptions` class acts as an intelligent configuration firewall.
By operating as a reactive, validating proxy, it shields developers from managing
manual synchronization loops and protects users from configuration errors, ensuring

> that options can safely change dynamically throughout a system's evolutionary life cycle.

Reviewer: ```suggestion
The :class:`_SolverOptions` class addresses these short comings.
By operating as a reactive, validating proxy, it shields developers from managing
manual synchronization loops and protects users from configuration errors, ensuring
that options can safely change dynamically throughout a system's evolutionary life cycle.
```

Core Features
-------------

The :class:`_SolverOptions` class inherits directly from Python's built-in
:repr:`dict`, behaving like a standard key-value map while injecting four

> crucial architectural features:

Reviewer: ```suggestion
The :class:`_SolverOptions` class inherits directly from Python's built-in
:repr:`dict`, behaving like a standard key-value map while adding four
important features:
```
